### ISSUE TITLE
How should I modify the loss function if I want to use soft labels for my action detection dataset?

### ISSUE THREAD

**Issue by jamesbond20181031** [Hope1337/YOWOv3#70]
hi，great work for me！
I have already tried several approaches:
Plan A： directly replaces the original gt-labels in the SimOTA class with gt-confs, where gt-confs denotes the soft-label probability distribution of the actions.
`
gt_bboxes, gt_labels,**gt_confs** = gt.split((4, nclass,nclass), 2)  # cls, xyxy
target_bboxes, target_scores, fg_mask = self.assign(scores, bboxes,
                                                            **gt_confs**, gt_bboxes, mask_gt,anchor_points * stride_tensor, stride_tensor)

changed from gt-labels to gt-confs

`

Plan B modifies the computation of target-scores inside the SimOTA class，in self.assign function.
original：（use true-labels）
`
target_labels.clamp(0)
target_scores = target_labels
`
now：（use true-confs as soft-labels）
`
if self.soft_label:
    target_scores = true_confs.view(-1, true_confs.shape[-1])[target_gt_idx]
else:
    target_scores=target_labels
`

Although both Plan A and Plan B can reduce the final loss to a certain level during training, the inference results are very poor—why is that?

**Comment by Hope1337** [comment]
Hi @jamesbond20181031 , I think the plan A is good enough. If, however, there are some problems about model's performance, I doubt that your dataset isn't qualitied as expected. Just my speculation, so I need some more context to inspect deeper.

**Comment by jamesbond20181031** [comment]
aditional,i load data with [-half-cliplen,half-cliplen],so in inference,i used centeridx of a clip as;
```center_frame_info = frame_buffer[clip_length // 2]```

however,in YOWOv3 forward function,you used the last frame as keyframe:
```key_frames = clips[:, :, -1, :, :]```
i didn`t change here yet.
does this matters?

**Comment by Hope1337** [comment]
@jamesbond20181031 I haven't looked at your code yet but I guess the key_frame choice is the problem. It causes a mismatch between training and inference phase as the detection code use the last frame to show the result.
You can verify that by experiment a little bit and in case it still doesn't work, ping me.

**Comment by jamesbond20181031** [comment]
> [@jamesbond20181031](https://github.com/jamesbond20181031) I haven't looked at your code yet but I guess the key_frame choice is the problem. It causes a mismatch between training and inference phase as the detection code use the last frame to show the result. You can verify that by experiment a little bit and in case it still doesn't work, ping me.

okay，i changed YOWOv3 forward function to maintain the match between training and inference phase and dataloading:
```
lenth=clips.shape[2]
key_frames = clips[:, :, lenth//2:lenth//2+1, :,:]
```
i ll experiment a little bit more

**Comment by jamesbond20181031** [comment]
i retrained 66epoch with both planA and planB,
the loss drop fig is :

<img width="2400" height="1500" alt="Image" src="https://github.com/user-attachments/assets/9cdf2d35-adf7-4d08-825a-995e45fd8b3c" />

the final trainlogs are :
<planA>
[DEBUG] loss_cls=0.012 loss_box=0.564 loss_dfl=0.282
[DEBUG] loss_cls=0.014 loss_box=0.843 loss_dfl=0.346
[DEBUG] loss_cls=0.012 loss_box=0.426 loss_dfl=0.293
[DEBUG] loss_cls=0.017 loss_box=0.460 loss_dfl=0.266
[DEBUG] loss_cls=0.011 loss_box=0.948 loss_dfl=0.313
[DEBUG] loss_cls=0.029 loss_box=1.926 loss_dfl=0.376
[DEBUG] loss_cls=0.013 loss_box=0.581 loss_dfl=0.259
[DEBUG] loss_cls=0.009 loss_box=0.447 loss_dfl=0.291
[DEBUG] loss_cls=0.016 loss_box=0.908 loss_dfl=0.279
[DEBUG] loss_cls=0.012 loss_box=0.680 loss_dfl=0.272
[DEBUG] loss_cls=0.015 loss_box=0.786 loss_dfl=0.290
[DEBUG] loss_cls=0.013 loss_box=0.798 loss_dfl=0.276
[DEBUG] loss_cls=0.011 loss_box=0.452 loss_dfl=0.301
epoch : 66, update : 92, loss = 1.1717647016048431

<planB>
[DEBUG] loss_cls=0.800 loss_box=0.714 loss_dfl=0.246
[DEBUG] loss_cls=0.858 loss_box=0.755 loss_dfl=0.362
[DEBUG] loss_cls=0.946 loss_box=1.264 loss_dfl=0.385
[DEBUG] loss_cls=0.661 loss_box=0.536 loss_dfl=0.266
[DEBUG] loss_cls=0.659 loss_box=0.469 loss_dfl=0.271
[DEBUG] loss_cls=0.991 loss_box=1.141 loss_dfl=0.300
[DEBUG] loss_cls=0.898 loss_box=0.938 loss_dfl=0.302
[DEBUG] loss_cls=0.679 loss_box=0.499 loss_dfl=0.301
[DEBUG] loss_cls=0.642 loss_box=1.000 loss_dfl=0.389
[DEBUG] loss_cls=0.890 loss_box=0.573 loss_dfl=0.435
epoch : 66, update : 92, loss = 2.010017476975918



using the infercode,the results are:
<planA>
正在处理以帧 144 为中心的片段...
Box coordinates range: [-3.681, 660.229]
Class probs range: [0.000, 0.011]
正在处理以帧 160 为中心的片段...
Box coordinates range: [2.969, 658.686]
Class probs range: [0.000, 0.001]
正在处理以帧 176 为中心的片段...
Box coordinates range: [0.168, 658.939]
Class probs range: [0.000, 0.005]
正在处理以帧 192 为中心的片段...
Box coordinates range: [-1.361, 666.557]
Class probs range: [0.000, 0.014]
正在处理以帧 208 为中心的片段...
Box coordinates range: [2.416, 669.889]
Class probs range: [0.000, 0.018]
正在处理以帧 224 为中心的片段...
Box coordinates range: [3.938, 656.828]
Class probs range: [0.000, 0.002]
正在处理以帧 240 为中心的片段...
Box coordinates range: [2.749, 659.610]
Class probs range: [0.000, 0.001]
正在处理以帧 256 为中心的片段...
Box coordinates range: [1.909, 662.008]
Class probs range: [0.000, 0.001]
正在处理以帧 272 为中心的片段...
Box coordinates range: [-12.946, 662.920]
Class probs range: [0.000, 0.001]
视频处理完成。

=== 检测统计 ===
总检测数: 0


<planB>
正在处理以帧 80 为中心的片段...
Box coordinates range: [6.559, 590.864]
Class probs range: [0.000, 0.004]
正在处理以帧 96 为中心的片段...
Box coordinates range: [7.152, 576.286]
Class probs range: [0.000, 0.004]
正在处理以帧 112 为中心的片段...
Box coordinates range: [7.042, 598.195]
Class probs range: [0.000, 0.004]
正在处理以帧 128 为中心的片段...
Box coordinates range: [7.313, 604.309]
Class probs range: [0.000, 0.004]
正在处理以帧 144 为中心的片段...
Box coordinates range: [7.265, 601.106]
Class probs range: [0.000, 0.004]
正在处理以帧 160 为中心的片段...
Box coordinates range: [6.997, 597.377]
Class probs range: [0.000, 0.005]
正在处理以帧 176 为中心的片段...
Box coordinates range: [7.082, 598.925]
Class probs range: [0.000, 0.005]
正在处理以帧 192 为中心的片段...
Box coordinates range: [7.147, 605.780]
Class probs range: [0.000, 0.004]
正在处理以帧 208 为中心的片段...
Box coordinates range: [7.150, 603.524]
Class probs range: [0.000, 0.004]
正在处理以帧 224 为中心的片段...
Box coordinates range: [7.419, 585.375]
Class probs range: [0.000, 0.004]
正在处理以帧 240 为中心的片段...
Box coordinates range: [6.716, 574.021]
Class probs range: [0.000, 0.004]
正在处理以帧 256 为中心的片段...
Box coordinates range: [6.271, 589.201]
Class probs range: [0.000, 0.004]
正在处理以帧 272 为中心的片段...
Box coordinates range: [6.134, 588.428]
Class probs range: [0.000, 0.004]
视频处理完成。

=== 检测统计 ===
总检测数: 0


both 2 plans detect nothing!

**Comment by jamesbond20181031** [comment]
i tried several loss-side experiments:  
**plan0 (50 epochs)**  
Replace the hard one-hot gt-label in the assignment with the soft gt_conf probabilities, i.e. use [0.98, 0, 0.02] instead of [1, 0, 0]; everything else stays the same.

****plan1 – plan3 keep the dynamic label assignment (SimOTA) and use hard labels to obtain a stable fg_mask.****  

**plan1 (50 epochs)**  
Feed both gt-labels and gt-confs into the assigner, but compute target_scores with gt_confs.

**plan2 (50 epochs)**  
Same as plan1 for target_scores, and change the cls loss to focal BCE loss.

**plan3 (80 epochs)**  
Same as plan1 for target_scores, but replace the cls loss with  
- VFL for positives: −tg * (1 − pos_scores)^γ * log(pos_scores)  
- VFL for negatives: − (neg_scores)^γ * log(1 − neg_scores)  
The rest follows SimOTA.

**plan4 (80 epochs)**  
Drop dynamic assignment completely; use static label assignment.  
- cls loss: focal BCE on both pos/neg samples  
- box loss: IoU loss  
- dfl loss: unchanged

The above five plans all converge normally during training, yet give very poor inference results — they failed completely.


<only plan5 and plan6 are training now>
**plan5**  
Use hard labels only for positive-sample assignment; let the soft labels penalise the cls-loss weight together with class_ratio weight.  
No weighting for negative samples.

**plan6**  
Same as plan5 for positives, but apply class_ratio-based weighting to negative samples as well (lower class_ratio → higher weight, forcing the model to pay more attention).


i guess plan6 would be my last chance, if it still can`t work well,i really don`t know how to do the Optimization and Improvement!god bless me!
